_posts/2022-10-05-pdf-generation-at-scale.md

```diff
@@ -0,0 +1,212 @@
+
+```bash
+mv page-to-pdf.yml stack.yml
+faas-cli new pdf-gen --lang node17 -f stack.yml
+```
+
+Edit the `stack.yaml` file to include the S3 secrets and bucket name. Note that we also set two additional environment variables `RAW_BODY` and `MAX_RAW_SIZE`. These are required to receive the generated PDFs on the `/save` endpoint.
+
+```yaml
+pdf-gen:
+    lang: node17
+    handler: ./pdf-gen
+    image: welteki2/pdf-gen:latest
+    secrets:
+      - s3-key
+      - s3-secret
+    environment:
+      s3_bucket: of-demo-pdf-gen
+      RAW_BODY: true
+      MAX_RAW_SIZE: 40Mb
+```
+
+Edit `gen-pdf/handler.js`
+
+```js
+'use strict'
+const { Upload } = require("@aws-sdk/lib-storage")
+const { S3Client } = require("@aws-sdk/client-s3")
+const axios = require('axios').default
+const fsPromises = require('fs').promises
+const crypto = require('crypto');
+
+let s3Client
+let bucketName = process.env.s3_bucket || "of-demo-pdf-gen"
+
+async function initS3() {
+  if (s3Client) return
+
+  let accessKeyId = await fsPromises.readFile('/var/openfaas/secrets/s3-key', 'utf8')
+  let secretAccessKey = await fsPromises.readFile('/var/openfaas/secrets/s3-secret', 'utf8')
+
+  s3Client = new S3Client({ region: "eu-central-1", credentials: { accessKeyId, secretAccessKey } })
+}
+
+module.exports = async (event, context) => {
+  await initS3()
+
+  if (event.path === '/save') { 
+    if (event.headers["content-type"] === "application/pdf") {
+      let uri = event.headers["x-page-uri"]
+
+      console.log(`Received pdf of: ${uri}`)
+
+      const hash = crypto.createHash('sha256').update(uri).digest('hex');
+      const key = `${hash}.pdf`
+
+      const  upload = new Upload({
+        client: s3Client,
+        params: {
+          Bucket: bucketName,
+          Key: key,
+          Body: event.body,
+          ContentType: "application/pdf",
+          Metadata: { uri }
+        },
+      })
+      await upload.done()
+      console.log(`Saved as: ${key}`)
+
+      return context.status(200).succeed()
+    }
+
+    return context.status(400).fail()
+  }
+
+  let body = JSON.parse(event.body)
+  let pages = body.pages
+  console.log("Submitting pages")
+
+  pages.forEach(async (uri) => {
+    const res = await axios.post(
+      "http://gateway.openfaas:8080/async-function/page-to-pdf",
+      { uri },
+      { headers: { 'X-Callback-Url': 'http://gateway.openfaas:8080/function/pdf-gen/save' } }
+    )
+  })
+
+  return context
+    .status(202)
+    .succeed()
+}
+```
+
+The function invokes `page-to-pdf` asynchronously and sets the `X-Callback-Url` header to receive the result once it has finished processing.
+
+## Conclusion
+You might also like:
+
+- [How to process your data the resilient way with back pressure](https://www.openfaas.com/blog/limits-and-backpressure/)
+- [Web scraping that just works with OpenFaaS with Puppeteer](https://www.openfaas.com/blog/puppeteer-scraping/)
```

Comment:
Let's also have this in the intro to explain how it was covered before, for folks who don't know how to scrape yet.